tetris.py

```diff
@@ -0,0 +1,175 @@
+                tt = [(n % self.x == self.x - 1) or (n + 1 in self.frozen) for n in self.current_configuration]
+                if not any(tt):
+                    for n in range(len(self.configurations[p])):
+                        self.configurations[p][n] += 1
+
+    def down(self):
+        if self.active:
+            for p in range(len(self.configurations)):
+                for n in range(len(self.configurations[p])):
+                    self.configurations[p][n] += self.x
+        # if active piece hit the floor or touched another, freeze it and update static set
+        tt = [((self.x * (self.y - 1)) <= n < (self.x * self.y)) or (n + self.x in self.frozen)
+              for n in self.current_configuration]
+        if any(tt):
+            self.frozen += self.current_configuration
+            self.configurations = [[]]
+            self.current_configuration = []
+            self.active = False
+        # if any column is completely filled: game over
+        for col in range(self.x):
+            tt = [(col + row * self.x) in (self.current_configuration + self.frozen) for row in range(self.y)]
+            if all(tt):
+                self.game_over = True
+
+    def line_clear(self):
+        # check for filled lines
+        for row in range(self.y):
+            tt = [n in self.frozen for n in range((row * self.x), ((row + 1) * self.x))]
+            # delete filled line and shift pieces above it down
+            if all(tt):
+                for p in range(len(self.frozen)):
+                    if self.frozen[p] < (row * self.x):
+                        self.frozen[p] += self.x
+                    elif (row * self.x) <= self.frozen[p] < ((row + 1) * self.x):
+                        # move deleted row elements under the edge of the grid :)
+                        self.frozen[p] = self.x * self.y + 1
+
+
+pieces = {'O': [[4, 14, 15, 5]],
+          'I': [[4, 14, 24, 34], [13, 14, 15, 16]],
+          'S': [[5, 4, 14, 13], [4, 14, 15, 25]],
+          'Z': [[4, 5, 15, 16], [5, 15, 14, 24]],
+          'L': [[4, 14, 24, 25], [5, 15, 14, 13], [4, 5, 15, 25], [6, 5, 4, 14]],
+          'J': [[5, 15, 25, 24], [15, 5, 4, 3], [5, 4, 14, 24], [4, 14, 15, 16]],
+          'T': [[4, 14, 24, 15], [4, 13, 14, 15], [4, 14, 24, 13], [13, 14, 15, 24]]}
+
+
+def get_dimensions():
+    dimensions = input("Board dimensions [x y]: ")
+    for char in dimensions:
+        if char not in "0123456789 ":
+            print("Dimensions should be integers!")
+            dimensions = get_dimensions()
+            print(dimensions)
+            break
+    return dimensions
+
+
+dim = get_dimensions().split()
+columns = int(dim[0]) if int(dim[0]) >= 10 else 10
+rows = int(dim[1])
+piece = random.choice("OISZLJT")
+
+grid = Grid(copy.deepcopy(pieces[piece]), columns, rows)
+grid.print_empty()
+grid.print()
+
+while True:
+    command = input("[rt=rotate, l=left, r=right, d=down, exit]: ")
+    if command == "rt":
```

Comment:
`grid.down()` call is used after every command except exit. It can be extracted to just one call.